### DISCUSSION TITLE
Login on new device notifications

### DISCUSSION THREAD

**Discussion by olivierboudet** [keycloak/keycloak#9390] [Ideas]
Hello,

I would like to send a mail notification to my users when a session is opened on a new device. Could this be implemented in Keycloak ?
I can try to implement something if you decide this is a good idea to have it in keycloak core.

PS : The JIRA [KEYCLOAK-4124](https://issues.redhat.com/browse/KEYCLOAK-4124) which propose the same thing was closed in 2019 as out of date.

**Comment by jbman** [comment]
The solution could be split into:

1. Emitting a dedicated event 'authentication on new device' with relevant data like IP and user agent. 
2. Handling this event by writing a mail to the user and/or store it at a built-in or custom user-related security history

To display information to the user, the user agent string must be analyzed so that browser and operating system can be shown. Project [yauaa](https://github.com/nielsbasjes/yauaa) does resolve info like device class, brand, operating system, etc.
Would this be part of 1 or 2?

For part 1 it also has to be discussed which info is used and stored to recognize a client as an existing one. Does anyone know of published best-practices how to do it?

**Reply by bs-matil** [reply]
If you have offline sessions you actually can detect a device for a kind of infinite time. 
Maybe some kind of unprivileged offline session which does not represent real access rights but rather the information that this is a known device would help. Like some kind of special client "device" which only exists to create this "device sessions" for this client on any device a user could have...

**Reply by jbman** [reply]
Yes, a device is identified as _new_ if it can't be recognized either by a information provided by the browser (i.e. user agent info or other fingerprinting techniques) or by a cookie which was placed there before by Keycloak. It could also be a combination. I would like to discuss and understand which information is extracted and stored client (browser) and server-side to identify a client.

A simple logic would be a cookie so that a device is always a new one when this cookie isn't send. This is probably the 'magic signed cookie on the device' @stianst talked about in the [KEYCLOAK-4124](https://issues.redhat.com/browse/KEYCLOAK-4124) comment. This also lists the idea to have new kind of sessions.

**Reply by olivierboudet** [reply]
I like the idea to have a new event when a new device is detected. I think the part 1 could be sufficient in a first step allowing users to write a SPI handling this event.
About the parsing of UserAgent to display information to the user, perhaps a new library is not needed as the class `DeviceActivityManager` seems to already do someting very similar ? Maybe a `DeviceRepresentation` could be added in the event ?
